File: sdks/go/insforge/http_client.go
package insforge

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// httpClient wraps net/http.Client with InsForge-specific behaviour.
type httpClient struct {
	baseURL     string
	anonKey     string
	accessToken string
	mu          sync.RWMutex
	headers     map[string]string
	client      *http.Client
}

func newHTTPClient(baseURL, anonKey string, headers map[string]string) *httpClient {
	return &httpClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		anonKey: anonKey,
		headers: headers,
		client:  &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *httpClient) setAccessToken(token string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.accessToken = token
}

func (c *httpClient) authToken() string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	if c.accessToken != "" {
		return c.accessToken
	}
	return c.anonKey
}

func (c *httpClient) buildRequest(ctx context.Context, method, path string, body interface{}, extraHeaders map[string]string) (*http.Request, error) {
	fullURL := c.baseURL + path
	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(data)
	}

	req, err := http.NewRequestWithContext(ctx, method, fullURL, bodyReader)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	if token := c.authToken(); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	for k, v := range c.headers {
		req.Header.Set(k, v)
	}
	for k, v := range extraHeaders {
		req.Header.Set(k, v)
	}
	return req, nil
}

func (c *httpClient) do(req *http.Request) (interface{}, error) {
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, &InsForgeError{Message: err.Error(), StatusCode: 0}
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &InsForgeError{Message: "failed to read response body", StatusCode: resp.StatusCode}
	}

	if resp.StatusCode == 204 || len(raw) == 0 {
		return nil, nil
	}

	var parsed interface{}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, &InsForgeError{Message: string(raw), StatusCode: resp.StatusCode}
	}

	if resp.StatusCode >= 400 {
		if m, ok := parsed.(map[string]interface{}); ok {
			return nil, errorFromBody(m, resp.StatusCode)
		}
		return nil, &InsForgeError{Message: string(raw), StatusCode: resp.StatusCode}
	}
	return parsed, nil
}

func (c *httpClient) get(ctx context.Context, path string, params url.Values, extraHeaders map[string]string) (interface{}, error) {
	fullPath := path
	if len(params) > 0 {
		fullPath += "?" + params.Encode()
	}
	req, err := c.buildRequest(ctx, http.MethodGet, fullPath, nil, extraHeaders)
	if err != nil {
		return nil, err
	}
	return c.do(req)
}

func (c *httpClient) post(ctx context.Context, path string, body interface{}, extraHeaders map[string]string) (interface{}, error) {
	req, err := c.buildRequest(ctx, http.MethodPost, path, body, extraHeaders)
	if err != nil {
		return nil, err
	}
	return c.do(req)
}

func (c *httpClient) put(ctx context.Context, path string, body interface{}, extraHeaders map[string]string) (interface{}, error) {
	req, err := c.buildRequest(ctx, http.MethodPut, path, body, extraHeaders)
	if err != nil {
		return nil, err
	}
	return c.do(req)
}

func (c *httpClient) patch(ctx context.Context, path string, body interface{}, extraHeaders map[string]string) (interface{}, error) {
	req, err := c.buildRequest(ctx, http.MethodPatch, path, body, extraHeaders)
	if err != nil {
		return nil, err
	}
	return c.do(req)
}

func (c *httpClient) delete(ctx context.Context, path string, params url.Values, extraHeaders map[string]string) (interface{}, error) {
	fullPath := path
	if len(params) > 0 {
		fullPath += "?" + params.Encode()
	}
	req, err := c.buildRequest(ctx, http.MethodDelete, fullPath, nil, extraHeaders)
	if err != nil {
		return nil, err
	}
	return c.do(req)
}

func (c *httpClient) uploadRaw(ctx context.Context, path string, data []byte, contentType string) (interface{}, error) {
	fullURL := c.baseURL + path
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, fullURL, bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", contentType)
	if token := c.authToken(); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	for k, v := range c.headers {
		req.Header.Set(k, v)
	}
	return c.do(req)
}

func (c *httpClient) downloadBytes(ctx context.Context, path string) ([]byte, error) {
	fullURL := c.baseURL + path
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fullURL, nil)
	if err != nil {
		return nil, err
	}
	if token := c.authToken(); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	for k, v := range c.headers {
		req.Header.Set(k, v)
	}
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, &InsForgeError{Message: err.Error()}
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, &InsForgeError{StatusCode: resp.StatusCode, Message: "download failed"}
	}
	return io.ReadAll(resp.Body)
}

File: sdks/python/insforge/lib/http_client.py
from __future__ import annotations

import json
from typing import Any

import httpx

from insforge.errors import InsForgeError


class HttpClient:
    def __init__(
        self,
        base_url: str,
        anon_key: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._anon_key = anon_key
        self._extra_headers = headers or {}
        self._access_token: str | None = None
        self._client = httpx.AsyncClient(timeout=30.0)

    def set_access_token(self, token: str | None) -> None:
        self._access_token = token

    def _auth_header(self) -> dict[str, str]:
        token = self._access_token or self._anon_key
        if token:
            return {"Authorization": f"Bearer {token}"}
        return {}

    def _build_headers(self, extra: dict[str, str] | None = None) -> dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            **self._extra_headers,
            **self._auth_header(),
        }
        if extra:
            headers.update(extra)
        return headers

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json_body: Any = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        content: bytes | None = None,
        content_type: str | None = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        req_headers = self._build_headers(headers)
        if content_type:
            req_headers["Content-Type"] = content_type
        if content is not None:
            req_headers.pop("Content-Type", None)

        response = await self._client.request(
            method,
            url,
            headers=req_headers,
            json=json_body,
            params=params,
            content=content,
        )

        if response.status_code >= 400:
            try:
                body = response.json()
            except Exception:
                body = {"error": {"message": response.text}}
            raise InsForgeError.from_response(body, response.status_code)

        if response.status_code == 204 or not response.content:
            return None

        return response.json()

    async def get(self, path: str, *, params: dict[str, Any] | None = None, headers: dict[str, str] | None = None) -> Any:
        return await self._request("GET", path, params=params, headers=headers)

    async def post(self, path: str, body: Any = None, *, params: dict[str, Any] | None = None, headers: dict[str, str] | None = None) -> Any:
        return await self._request("POST", path, json_body=body, params=params, headers=headers)

    async def put(self, path: str, body: Any = None, *, headers: dict[str, str] | None = None) -> Any:
        return await self._request("PUT", path, json_body=body, headers=headers)

    async def patch(self, path: str, body: Any = None, *, headers: dict[str, str] | None = None) -> Any:
        return await self._request("PATCH", path, json_body=body, headers=headers)

    async def delete(self, path: str, *, params: dict[str, Any] | None = None, headers: dict[str, str] | None = None) -> Any:
        return await self._request("DELETE", path, params=params, headers=headers)

    async def upload(self, path: str, data: bytes, content_type: str, *, headers: dict[str, str] | None = None) -> Any:
        req_headers = self._build_headers(headers)
        req_headers["Content-Type"] = content_type
        response = await self._client.put(
            f"{self._base_url}{path}",
            content=data,
            headers=req_headers,
        )
        if response.status_code >= 400:
            try:
                body = response.json()
            except Exception:
                body = {"error": {"message": response.text}}
            raise InsForgeError.from_response(body, response.status_code)
        if not response.content:
            return None
        return response.json()

    async def download_bytes(self, path: str) -> bytes:
        url = f"{self._base_url}{path}"
        response = await self._client.get(url, headers=self._build_headers())
        if response.status_code >= 400:
            try:
                body = response.json()
            except Exception:
                body = {"error": {"message": response.text}}
            raise InsForgeError.from_response(body, response.status_code)
        return response.content

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> HttpClient:
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.aclose()
